### PR TITLE
StoreFactory cleanup

### PR DESCRIPTION
**Pull request by chrisvest**
Splits NeoStore into NeoStores and MetaDataStore, where NeoStores is about holding references to the other stores, and making sure they are initialised. The responsibility of creating store files is pushed down, into the stores themselves.
- This introduces the NeoStores concept, as the class responsible for holding
  references to the other stores.
- The NeoStores is also in charge of initialising the other stores, and does so
  lazily, so tools can continue to work on incomplete store directories.
- The stores themselves are now in charge of creating their store files, and
  writing headers and trailers to them.
- The StoreFactory has been reduced to something that gathers up configuration
  information, and then creates the NeoStores.
- NeoStore has been renamed to MetaDataStore, though the file itself is still
  called "neostore".
- The responsibility of closing stores has also been migrated from the
  individual stores, into the NeoStores.

### CODE REVIEW

**Comment by thobe** [comment]
I would like to see these store classes named differently.

Given that all filenames start with `neostore`, having `NeoStore` as the store that stores metadata makes little sense, I'd rather call that class the `MetadataStore`, and then call the aggregating class `NeoStores`.

That is of course just my suggestion, and I'm open to others, but I think the names proposed by this PR are confusing, since the `MetaStore` in this implementation does not store any meta data, and the `NeoStore` does store metadata... while all store file names contain the prefix `neostore`.

**Comment by chrisvest** [comment]
@thobe I have updated the PR with the names you suggest.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/store/DumpStore.java [hunk 1]
```diff
@@ -95,34 +106,37 @@ else if ( !file.isDirectory() && file.getName().indexOf( ':' ) != -1 )
                 throw new IllegalArgumentException( "No such file: " + arg );
             }
         }
-        switch ( file.getName() )
+        try ( NeoStores neoStores = createStoreFactory.apply( file ).openNeoStores( false ) )
```
Comment (thobe) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/DumpStore.java:110]: This looks like this now requires all store files. Previously we were able to dump the content of a single store file without needing to have the other store files present.

Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/DumpStore.java:110]: NeoStores opens the stores lazily as they are requested, so it looks like we open all of them, but it's not actually the case.

Comment (thobe) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/store/DumpStore.java:110]: Yes, confirmed, cool.

**Comment by thobe** [comment]
Looks good to me.

**Comment by davidegrohmann** [comment]
@chrisvest this needs a rebase

**Pull request merged**